### PR TITLE
Fix 403 error with downloading images from Digikey's CDN provider Akamai

### LINKED ISSUE #5: Failure to download image from Digikey
I have recently setup inventree and found this very helpful tool.  Been playing with it today and all is working great, except downloading of images from digikey.  I see the following error, but on inspection I see that it's failing to download the image.  It's getting a 403 error code.  I have downloaded the image from a browser with no issues.  I thought maybe I had some cached authentication in the browser cache.  So I tried a browser that I've not used to log into digikey and it also downloaded just fine.  So a bit puzzled why the script would fail with a 403.  

```
Trying URL https://mm.digikey.com/Volume0/opasdata/d220001/medias/images/5565/MFG_PI_5_HERO.jpg
<Response [403]>
ERROR: Request code is 403
Traceback (most recent call last):
  File "/usr/local/Cellar/python@3.9/3.9.16/Frameworks/Python.framework/Versions/3.9/lib/python3.9/runpy.py", line 197, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/local/Cellar/python@3.9/3.9.16/Frameworks/Python.framework/Versions/3.9/lib/python3.9/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/Users/bpayne/wip/inventory/Digikey-Inventree-Integration/inventree_digikey/__main__.py", line 23, in <module>
    add_digikey_part(dkpart)
  File "/Users/bpayne/wip/inventory/Digikey-Inventree-Integration/inventree_digikey/Inventree.py", line 27, in add_digikey_part
    inv_part = create_inventree_part(dkpart)
  File "/Users/bpayne/wip/inventory/Digikey-Inventree-Integration/inventree_digikey/Inventree.py", line 81, in create_inventree_part
    upload_picture(dkpart, part)
  File "/Users/bpayne/wip/inventory/Digikey-Inventree-Integration/inventree_digikey/Inventree.py", line 121, in upload_picture
    invPart.uploadImage(img_file)
  File "/Users/bpayne/wip/inventory/venv/lib/python3.9/site-packages/inventree/base.py", line 543, in uploadImage
    raise TypeError(f"uploadImage called with invalid image: '{image}'")
TypeError: uploadImage called with invalid image: '-1'
```

### PR DESCRIPTION
**Pull request by benpayne**
The requests api is being blocked by Akamai and returning 403 errors.  Loading these same URLs in a browser works just fine and wget will download them also.  Curl will not, at first I thought HTTP headers, but the http.client API works with not header modifications at all.  I would guess it's something about the SSL library used by these tools, but that's just a guess.

TLDR - Changes to using http.client instead of requests api and downloading images now works without 403 errors

Closes #5

### CODE REVIEW

**Comment by EUdds** [comment]
Hey, thanks for contributing! I'm gonna test this out and see how it goes!!

**Comment by EUdds** [comment]
This PR may prompt me to make some real tests for this package. Overall running this on a few parts on my local instance works well, but I have not fully validated. Gonna try to spin up a branch tn or sometime early this week with some pytests. Thanks again for your contribution! It's cool to see people using it

**Comment by EUdds** [comment]
@benpayne I got some CI and unit tests kicked off. It created some conflicts that I'm happy to take care of for you if you'd like. But with some heavy (and much-needed and appreciated) changes coming, it was a good time to kick off some real testing. Let me know!

Thanks again!

**Comment by benpayne** [comment]
Agreed, I'm going to rebase these changes and the other changes I've been making.  Tests are great.  You mentioned have other changes coming?  Is there some things you've working on?  Not sure if you took a look at my repo to see what I've been doing.  But would be good to compare notes.

**Comment by EUdds** [comment]
> Agreed, I'm going to rebase these changes and the other changes I've been making. Tests are great. You mentioned have other changes coming? Is there some things you've working on? Not sure if you took a look at my repo to see what I've been doing. But would be good to compare notes.

> But with some heavy (and much-needed and appreciated) changes coming

Sorry, I was referencing your incoming changes in this PR here :)

File: inventree_digikey/Inventree.py [hunk 1]
```diff
@@ -19,16 +21,51 @@
...
 def load_config():
     config = configparser.ConfigParser()
     config.read(CONFIG_FILE_PATH)
     global API_URL, USERNAME, PASSWORD, API
+    global catagory_map, params_map, htsus_pk
 
     API_URL = config['INVENTREE_API']['URL']
     USERNAME = config['INVENTREE_API']['USER']
     PASSWORD = config['INVENTREE_API']['PASSWORD']
     API = InvenTreeAPI(API_URL, username=USERNAME, password=PASSWORD)
```
Comment (EUdds) [inline inventree_digikey/Inventree.py:37]: I'll push this off into another PR but I like the idea of making the config file it's own module. Python global variables are kinda icky and I'm not a huge fan of how I did this. But having one module to read the ini file and map to variable locations could make extending the config.ini easier in the future

File: inventree_digikey/Inventree.py [hunk 2]
```diff
@@ -19,16 +21,51 @@
...
+    inventree_params = ParameterTemplate.list(API)
+    #for param in inventree_params:
+    #    print(f"{param.pk} : {param.name}")
+
+    for param in config['PARAMETERS']:
+        value = config['PARAMETERS'][param]
+        match_param = [p.pk for p in inventree_params if p.name == value]
+        if len(match_param) > 0:
+            #print(f"Found {value} in inventree with pk {match_param[0]}")
+            params_map[param] = match_param[0]
```
Comment (EUdds) [inline inventree_digikey/Inventree.py:48]:
```suggestion
        if len(match_param) == 1:
            #print(f"Found {value} in inventree with pk {match_param[0]}")
            params_map[param] = match_param[0]
        elif len(match_param) > 1:
            raise ValueError(f"Found more than one possible match for parameter {param}")
```

File: inventree_digikey/Inventree.py [hunk 3]
```diff
@@ -19,16 +21,51 @@
...
+    inventree_categories = PartCategory.list(API)
+    #for cat in inventree_categories:
+    #    print(f"{cat.pk} : {cat.name}")
+
+    for cat in config['CATEGORIES']:
+        value = config['CATEGORIES'][cat]
+        match_cat = [c.pk for c in inventree_categories if c.name == value]
+        if len(match_cat) > 0:
+            #print(f"Found {value} in inventree with pk {match_cat[0]}")
+            catagory_map[cat] = match_cat[0]
```
Comment (EUdds) [inline inventree_digikey/Inventree.py:63]: Same suggestion as above, maybe abstract this searching logic out into a function since it's repeated from above.

File: inventree_digikey/Inventree.py [hunk 4]
```diff
@@ -86,27 +122,32 @@ def create_inventree_part(dkpart: DigiPart):
...
-    print("="*20)
-    print("Choose a category")
-    for idx, category in enumerate(categories):
-        print("\t%d %s" %(idx, category.name))
-    print("="*20)
-    idx = int(input("> "))
-    return categories[idx]
+    if str.lower(dkpart.category) in catagory_map:
+        return catagory_map[str.lower(dkpart.category)]
+    else:
```
Comment (EUdds) [inline inventree_digikey/Inventree.py:141]:
```suggestion
    clean_category = str.lower(dkpart.category)
    if clean_category in catagory_map:
        return catagory_map[clean_category]
    else:
```
This is a nitpick, not required if you don't want to

**Review by EUdds — CHANGES_REQUESTED** [review]
Looks great! Minor comments. I think I will get started on that config module (I'll start off of your changes so you don't have to change what you just wrote)

File: inventree_digikey/Inventree.py (hunk 1 above)
Comment (benpayne) [reply inventree_digikey/Inventree.py:37]: I think the idea of the config modules is great.  I agree about globals == Bad.  Was just looking to minimize change at this time and followed your lead.  But looked over the config changes and that looks great.

File: inventree_digikey/Inventree.py (hunk 4 above)
Comment (benpayne) [reply inventree_digikey/Inventree.py:143]: Good catch.

File: inventree_digikey/Inventree.py (hunk 3 above)
Comment (EUdds) [reply inventree_digikey/Inventree.py:65]: Can you add the changes from the parameter mapping here too?

**Review by EUdds — COMMENTED** [review]
It looks good, could you add the if match greater than 1 exception here too? Also please remove commented-out print functions

File: inventree_digikey/Digikey.py [hunk 5]
```diff
@@ -1,15 +1,34 @@
...
 
+config = configparser.ConfigParser()
+config_path = Path(__file__).resolve().parent / 'config.ini'
+config.read(config_path)
 
-def get_part_from_part_number(partnum: str):
+os.environ['DIGIKEY_CLIENT_ID'] = config['DIGIKEY_API']['CLIENT_ID']
+os.environ['DIGIKEY_CLIENT_SECRET'] = config['DIGIKEY_API']['CLIENT_SECRET']
+os.environ['DIGIKEY_CLIENT_SANDBOX'] = 'False'
+os.environ['DIGIKEY_STORAGE_PATH'] = '.'
```
Comment (EUdds) [inline inventree_digikey/Digikey.py:15]: Moving this into the uppermost scope of the script is what's failing testing

File: inventree_digikey/__main__.py [hunk 6]
```diff
@@ -1,23 +1,28 @@
...
-config_path = Path(__file__).resolve().parent / 'config.ini'
-config.read(config_path)
+# Add an optional '-y' flag to bypass prompting
+parser.add_argument('-y', action='store_true', help='Bypass user prompts and assume "yes"')
+parser.add_argument('-o', action='store_true', help='Query and Order number and import it')
 
-os.environ['DIGIKEY_CLIENT_ID'] = config['DIGIKEY_API']['CLIENT_ID']
-os.environ['DIGIKEY_CLIENT_SECRET'] = config['DIGIKEY_API']['CLIENT_SECRET']
-os.environ['DIGIKEY_CLIENT_SANDBOX'] = 'False'
-os.environ['DIGIKEY_STORAGE_PATH'] = '.'
```
Comment (EUdds) [inline inventree_digikey/__main__.py:15]: Please leave the configuration stuff here until the config reader stuff is done

**Comment by benpayne** [comment]
So there are a lot of things about this over all code that need to be cleaned up.  You're pointing out a bunch of them.  I think I should take a pass on cleaning this up and resubmit the PR.  I'm also working on some other features too.  I'm going to start a discussion thread about all this so I can get your input on my ideas.  I'm sure you've had some ideas of things to add as well.  So I'm going to close this for now.
